lib/rgeo/impl_helper/valid_op.rb

```diff
@@ -0,0 +1,373 @@
+# frozen_string_literal: true
+
+require "set"
+
+module RGeo
+  module ImplHelper
+    module TopologyErrors
+      # Standard error messages from
+      # https://github.com/locationtech/jts/blob/0afbfb1956ec24912a8b4dc4edff0f1200442857/modules/core/src/main/java/org/locationtech/jts/operation/valid/TopologyValidationError.java#L98-L110
+      TOPOLOGY_VALIDATION_ERR = "Topology Validation Error"
+      REPEATED_POINT = "Repeated Point"
+      HOLE_OUTSIDE_SHELL = "Hole lies outside shell"
+      NESTED_HOLES = "Holes are nested"
+      DISCONNECTED_INTERIOR = "Interior is disconnected"
+      SELF_INTERSECTION = "Self-intersection"
+      RING_SELF_INTERSECTION = "Ring Self-intersection"
+      NESTED_SHELLS = "Nested shells"
+      DUPLICATE_RINGS = "Duplicate Rings"
+      TOO_FEW_POINTS = "Too few distinct points in geometry component"
+      INVALID_COORDINATE = "Invalid Coordinate"
+      UNCLOSED_RING = "Ring is not closed"
+    end
+
+    # Mixin based off of the JTS/GEOS IsValidOp class.
+    # Implements #valid? and #invalid_reason on Features that include this.
+    #
+    # @see https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/operation/valid/IsValidOp.java
+    module ValidOp
+      # TODO: determine if "Unkown Validity" should count as valid.
```

Comment:
I'd say it shouldn't, do you have examples @keithdoggett ?

Reply:
I agree that it shouldn't. If an error is raised along the way there is probably some sort of issue with the geometry anyways.